pkg/workload/sysbench.go

```diff
@@ -0,0 +1,648 @@
+	e[TypePointSelect] = s.eventPointSelect
+	e[TypeUpdateIndex] = s.eventIndexUpdate
+	e[TypeUpdateNonIndex] = s.eventNonIndexUpdate
+	e[TypeReadWrite] = s.eventReadWrite
+	s.events = e
+
+	h := make(map[string]func(ctx context.Context, w *sysbenchWorker) error)
+	h[QueryPointSelect] = s.execPointSelect
+	h[QueryIndexUpdate] = s.execIndexUpdate
+	h[QueryNonIndexUpdate] = s.execNonIndexUpdate
+	h[QuerySimpleRanges] = s.execSimpleRanges
+	h[QuerySumRanges] = s.execSumRanges
+	h[QuerySimpleRanges] = s.execSimpleRanges
+	h[QuerySumRanges] = s.execSumRanges
+	h[QueryOrderRagnes] = s.execOrderRanges
+	h[QueryDistinctRanges] = s.execDistinctRanges
+	// h[QueryDeletes] = s.execDeletes
+	// h[QueryInserts] = s.execInserts
+	s.handlers = h
+
+	op := make(map[string][]string)
+	op[TypePointSelect] = make([]string, 0)
+	op[TypePointSelect] = append(op[TypePointSelect], QueryPointSelect)
+	op[TypeUpdateIndex] = make([]string, 0)
+	op[TypeUpdateIndex] = append(op[TypeUpdateIndex], QueryIndexUpdate)
+	op[TypeUpdateNonIndex] = make([]string, 0)
+	op[TypeUpdateNonIndex] = append(op[TypeUpdateNonIndex], QueryNonIndexUpdate)
+	op[TypeReadWrite] = make([]string, 0)
+	op[TypeReadWrite] = append(op[TypeReadWrite], QueryPointSelect,
+		QuerySimpleRanges, QuerySumRanges, QueryOrderRagnes, QueryDistinctRanges,
+		QueryIndexUpdate, QueryNonIndexUpdate,
+		QueryDeletes, QueryInserts,
+	)
+	s.operations = op
+
+	return nil
+}
+
+func (s *sysBench) Exec(ctx context.Context, tid int) {
+	cmdType := s.p.GetString(prop.SysbenchCmdType, "nil")
+	wlType := s.p.GetString(prop.SysbenchWorkLoadType, "nil")
+
+	switch cmdType {
+	case "prepare":
+		s.Prepare(tid)
+	case "run":
+		s.RunEvent(ctx, tid, wlType)
+	case "cleanup":
+		s.Cleanup(tid)
+	}
+}
+
+type sysbenchWorker struct {
+	r    *rand.Rand
+	conn *sql.Conn
+	tid  int
+	//map[Queryxxx]:[nil,table1_stmt,table2_stmt]
+	stmts map[string][]*sql.Stmt
+}
+
+func (s *sysBench) createWorker(ctx context.Context, tid int, wlType string) *sysbenchWorker {
+	conn, err := s.db.ToSqlDB().Conn(ctx)
+	if err != nil {
+		panic(err)
+	}
+	w := new(sysbenchWorker)
+	w.r = rand.New(rand.NewSource(time.Now().UnixNano() + int64(tid)*int64(100000000)))
+	w.conn = conn
+	w.tid = tid
+	w.stmts = s.prepareStatements(ctx, wlType, conn)
+
+	return w
+}
+
+func (s *sysBench) releaseWorker(ctx context.Context, w *sysbenchWorker) {
+	s.closeStatements(ctx, w.stmts)
+	w.conn.Close()
+}
+
+func (s *sysBench) RunEvent(ctx context.Context, tid int, wlType string) {
+	var err error
+	w := s.createWorker(ctx, tid, wlType)
+	defer s.releaseWorker(ctx, w)
+	event, ok := s.events[wlType]
+	timeStart := time.Now()
+
+	if !ok {
+		fmt.Println("sysbench workload type doesn't not exist")
+		return
+	}
+
+	for i := int64(0); i < s.eventCnt; i++ {
+		start := time.Now()
+		err = event(ctx, w)
+
+		if err != nil {
+			//fmt.Println(err)
+			return
+		}
+
+		sysBenchMeasure(start, wlType, err)
+
+		dur := time.Since(timeStart)
+		if dur >= s.time {
```

Comment:
We can create a context with timeout and pass it to functions so that the command can quit as soon as time over.